### ISSUE TITLE
Structure of linuxkit repository

### ISSUE THREAD

**Issue by deitch** [linuxkit/linuxkit#2122]
As raised in #2120, changing a core package requires:

1. Changing the package itself, e.g. `pkg/sshd/` or `pkg/getty/` or `pkg/init/`
2. Changing the many samples `.yml` files in `examples/` and in many of the `projects/` to use the updated image (changed hashes).

The current path has the following issues:

* Changes requested as part of the PR process create a new hash, which in turn, requires changes to the hashes in the example and projects `.yml` files
* The core changes might be to 3-4 files in the `pkg/` dir, but it gets lost in the 20 or more single line changes to `.yml` files
* Rebasing can get messy as multiple changes go through while a PR is open (quote "it is already hard enough to get one such PR merged with racing against other changes and rebasing (and therefore regenerating hashes etc")
* Permissions around maintenance can become messy as this grows (although notary makes it somewhat easier)

The question raised on that PR and migrated here is as follows. Is our current structure optimal? Some possibilities:

1. Move each `pkg/` into its own GitHub repo, plus one for examples and one for projects (or perhaps one for each project), leaving `linuxkit/linuxkit` as primarily for the tool and docs on how to use it.
2. Move each `pkg/` into its own repo, but keep examples and/or projects as part of `linuxkit/linuxkit` 
3. Keep it in a single repo, but enforce a two-step PR rule: one PR for the core change, a second PR to use that change in projects and examples
4. Keep it as is as the lesser of many evils
5. Other ideas?

/cc @dave-tucker @rn @justincormack @ijc @riyazdf

**Comment by dave-tucker** [comment]
One solution to this might lie in https://github.com/moby/tool/pull/96
We could keep the same structure and have a top level `latest.yml` checked in.
It's this file that gets updated when a package is changed.

This means that you could, for example, `moby build examples/sshd.yml latest.yml` and through the power of substitution be using the latest version of each package.

**Comment by deitch** [comment]
Is that the extensibility proposal, same as `docker-compose -f one.yml -f two.yml` etc., where the third overrides the second overrides the first?

As a general direction of moby composability, I like it. I still think we should have a real discussion about whether the 4 major parts - core tool, multiple `pkg/`, examples and projects - should be in a single repo.

**Comment by rn** [comment]
As commented on the referenced PR, I think a sensible (and working) approach is to use a single PR with multiple commits, with the last commit updating the YAML files to new hashes. This makes it trivial to rebase should it be necessary.

The current dance for folks who don't have signing keys set up for packages whereby someone needs to push/sign a package for CI to pass will go away once the CI is capable of building/pushing/signing (though this is still somewhat off, timewise).

https://github.com/linuxkit/linuxkit/pull/2117 adds a new make target to print out the current hash, which makes it easier to update YAML files.

So overall, I think the current model (potentially with additional support from the `moby` tool to re-use some of the boilerplate) is pretty workable. I don't like the idea of more repos and multiple PRs

**Comment by deitch** [comment]
> This makes it trivial to rebase should it be necessary.

Not really. Let's say we are working on three different `pkg/` changes in parallel, each of which leads to new hashes for the `examples/`. Sure, they are different lines, so in theory git should handle them sanely without requiring rebase, but wouldn't you _want_ to get the basic changes in?

`pkg/init/` in one PR definitely won't require rebasing because of changes to `pkg/sshd/` in another PR, but `examples/` in one PR very likely will because of changes to `examples/` in another.

> #2117 adds a new make target to print out the current hash, which makes it easier to update YAML files.

That is funny. In most of my PRs, I have added a `make hash` target that spits out the hash for exactly that reason... and got asked to remove it to hew to the standard. :-)

**Comment by justincormack** [comment]
Once we have packages being built in CI, so we know there will always be a new package if something is updated, then it becomes easier. As we actually know what the hash should be, we could generate the examples. We should definitely write a script that will update them to what the hashes should be.

**Comment by ijc** [comment]
Perhaps we should generate most of the `yml` files (toplevel and examples as a starter) at build time by substituting in the latest hashes into a template?

I think GH supports suppressing diffs to certain files based on gitattributes or something, but I'm reluctant to suggest using that to suppress diffs of `*.yml` since that would hide other real changes.

**Comment by justincormack** [comment]
Also we are adding https://github.com/moby/tool/pull/96 ie allowing substitutions to the image names, in theory we could have the files without tags and a substitution map that turns untagged (ie `latest`) into the actual current tag, and always use that map for the tests.

**Comment by deitch** [comment]
> we could generate the examples

Doesn't that assume we always want the latest?

**Comment by justincormack** [comment]
We always want latest for `test/` and `examples/` and `blueprints/` I think. Not for projects.

**Comment by deitch** [comment]
There is a lot of boilerplate as it is. The idea of a top-level `linuxkit.yml` followed by overrides for specifics might resolve both of these issues.

Even so, that assumes a lot of overlap. If there isn't so much, we will have repetition.

Coming back to the core issue raised: is there no interest in splitting `pkg/` out into separate repos? I would have thought that would be most modular for mgmt?

**Comment by dave-tucker** [comment]
The files could also include tags which would indicate a "blessed" configuration. We could add tests for both. Then it would point it if either the blessed config or the latest config were broken.

**Comment by deitch** [comment]
@dave-tucker I didn't understand. blessed?

**Comment by justincormack** [comment]
I think splitting out each `pkg` into repos is going to be too annoying in the short term, at least while we build them with `Makefile`s which are currently now mostly shared. If we have a build tool that just knows how to build and tag them it might be more manageable.
